### DISCUSSION TITLE
How can I get Star Wars Dark Forces II to work with dd7to9?

### DISCUSSION THREAD

**Discussion by eddmanx** [elishacloud/dxwrapper#231] [General]
It is the original release of the game, not GOG.
The guide says it is fully supported, but only launches to a black screen with white borders and then terminates.

I copied over ddraw.dll, dxwrapper.dll and dxwrapper.ini and set dd7to9 to 1. I had set the game to software mode beforehand.
I also tried a few fix options mentioned on the dd7to9 wiki page but no change.

**Comment by elishacloud** [comment]
Yes, you need to disable 3D acceleration in the in-game options before the game will work.  Also, I am using the GOG version.  Not sure if there is another issue with the CD version.

You could try setting the `FullscreenWindowMode` or `DdrawDisableDirect3DCaps` options and see if either of these help.

**Reply by eddmanx** [reply]
FullscreenWindowMode just reminded me of exclusive fullscreen, and that I had set a custom resolution at 48 Hz. Changing it to the usual 50, 60 or 75 makes it work. No extra wrapper settings are needed either, just dd7to9 enabled.

I guess the wrapper doesn't automatically handle refresh rate change? With dgvoodoo, the game changes the refresh rate to 50 Hz.

**Reply by elishacloud** [reply]
dxwrapper will use [whatever refresh rate](https://github.com/elishacloud/dxwrapper/blob/81c8291efa381207fe4cc8e2aba515a1d315ee07/ddraw/IDirectDrawX.cpp#L2909-L2910) chosen by the app.

Note: in windowed mode it uses the refresh rate of the desktop.   But keep in mind that some games only send X number of frames to the screen every second.  So even if the refresh rate is higher some of the frames will be duplicate, so it will still only show that many new frames.

**Reply by eddmanx** [reply]
At 48 Hz it crashes both with or without FullscreenWindowMode though.

I checked the game without any wrapper and it changes the refresh rate from 48 to 75 Hz and I can get into game.
For some reason dxwrapper crashes the game when that happens.

**Reply by elishacloud** [reply]
Please attach a log file when the game crashes here from the refresh rate changes.

Note: keep in mind that dgVoodoo2 is emulating DirectDraw and then outputting it in DirectX11/12.  Whereas dxwrapper is attempting to do a direct translation function call by function call from DirectDraw to Direct3D9.

**Reply by eddmanx** [reply]
[dxwrapper-jk.log](https://github.com/elishacloud/dxwrapper/files/13046363/dxwrapper-jk.log)

**Reply by elishacloud** [reply]
Nothing in that log.  Can you use the debug version of dxwrapper and send me the log file of the crash with the debug version?

**Reply by eddmanx** [reply]
[dxwrapper-jk.log](https://github.com/elishacloud/dxwrapper/files/13046629/dxwrapper-jk.log)

**Reply by eddmanx** [reply]
I thought what would happen if I applied dgvoodoo on top of dd7to9 (d3d9 to 11), and the game launches, however, the image occupies only a small part of the screen at the top-left corner.

EDIT: It only goes as far as the menu though. The only resolution listed is 1920x1080. When I try to start the campaign, it throws an error about not enough memory, and then another one about not being able to set the previous display mode (probably because of only 1920x1080 being available).

**Reply by elishacloud** [reply]
From the log it looks like the game never even sets the refresh rate.  The game calls `EnumAdapterModes()` twice and then exits.  I guess it does not find the resolution and/or refresh rate it is looking for.  I suspect that your video card does not have any 48 or 50 Hz resolutions and thus the game does not find any resolution for this call and just exits.

I need to think about the right fix for this.  But I guess this fix might help:  [dxwrapper.zip](https://github.com/elishacloud/dxwrapper/files/13047361/dxwrapper.zip)

**Reply by elishacloud** [reply]
> The only resolution listed is 1920x1080. When I try to start the campaign, it throws an error about not enough memory

Yeah, the game has a bug here.  I don't believe that is related to dxwrapper.  You need to set the game to a lower resolution.  For testing try: 1024x768

**Comment by eddmanx** [comment]
> I suspect that your video card does not have any 48 or 50 Hz resolutions and thus the game does not find any resolution for this call and just exits.

48 Hz is a custom resolution I made, but 50 Hz does exist by default, alongside 59, 60 and 75. It's odd that dxwrapper's presence somehow prevents the game from finding those when I'm using 48 Hz (although it might be the case with any custom Hz, not just 48. EDIT: Yea, I created a 72 Hz one and it's the same; the game with dxwrapper can't find the regular refresh rates).

> You need to set the game to a lower resolution. For testing try: 1024x768

In that test scenario (48 Hz display and dxwrapper + dgvoodoo), only 1920x1080 is listed in the game settings, so I can't pick any other resolution.

EDIT: The new dll crashes too.

**Reply by elishacloud** [reply]
Ok, here is the debug version.  Send me the log file if this does not work: [dxwrapper.zip](https://github.com/elishacloud/dxwrapper/files/13047786/dxwrapper.zip)

> (48 Hz display and dxwrapper + dgvoodoo), only 1920x1080 is listed in the game settings

Could be related to the resolutions that dgvoodoo gives out.  In this case dxwrapper is getting the resolutions from dgvoodoo rather than Windows.

**Reply by eddmanx** [reply]
Heeey, it works. The game now launches and picks 75 Hz (EDIT: using it only, without dgvoodoo). Two things I noticed though:
1. Resolutions are duplicated, for example, four 800x600 are listed. Also the max available res is 1152x864.
2. Sometimes after quitting, the jk.exe process fails to terminate and gets stuck.
[dxwrapper-jk.zip](https://github.com/elishacloud/dxwrapper/files/13048073/dxwrapper-jk.zip)

**Reply by elishacloud** [reply]
> 1. Resolutions are duplicated, for example, four 800x600 are listed. Also the max available res is 1152x864.

Yes, this is a byproduct of the way I fixed it.  I'll have to put a better fix below.

> 2. Sometimes after quitting, the jk.exe process fails to terminate and gets stuck.

Not sure about this one.  It could be because of the debug version?  It should not related to the fix I added.  You could also try setting `SingleProcAffinity`.  Some older games don't work well with multiple processors.

Here is a fix.  This should fix at least issue 1.: [dxwrapper.zip](https://github.com/elishacloud/dxwrapper/files/13048282/dxwrapper.zip)

**Reply by eddmanx** [reply]
It behaves the same.

**Reply by elishacloud** [reply]
Are you still getting duplicate resolutions?

**Reply by eddmanx** [reply]
Yes, I don't see any difference between this and the previous dll.

**Reply by eddmanx** [reply]
Just noticed the code was updated with a duplicate fix. I downloaded the binaries from Actions. No more duplicates, and it goes up to 1920x1080 as expected.

**Reply by elishacloud** [reply]
Yeah, I must have posted the wrong update.  Glad it is working for you now!

**Comment by eddmanx** [comment]
~~Testing the game on XP in VMware, the rendered image is not fullscreen but the UI is. I tried dxwrapper but it didn't change anything.~~

~~Any way it can be fixed?~~

Alright, false alarm; that's just the view size in the game's settings. :D

![jkdf2-dxwrapper-XP](https://github.com/elishacloud/dxwrapper/assets/13473205/64b5a858-db37-4f48-84f2-8afcb1811f5c)

**Reply by elishacloud** [reply]
Ok, great.  I was wondering about that!
